File: src/services/theme.ts
/**
 * Theme service - handles theme persistence using localStorage
 * 
 * This replaces the old Tauri-based theme service.
 */

import type { ThemeMode, ThemeName } from '../types/theme';

const THEME_STORAGE_KEY = 'rainy-day-theme';

interface StoredTheme {
  mode: ThemeMode;
  name: ThemeName;
}

/**
 * Get the current theme from localStorage
 */
export async function getTheme(): Promise<{ mode: ThemeMode; name: ThemeName }> {
  try {
    const stored = localStorage.getItem(THEME_STORAGE_KEY);
    if (stored) {
      const parsed = JSON.parse(stored) as StoredTheme;
      return {
        mode: parsed.mode || 'automatic',
        name: parsed.name || 'default',
      };
    }
  } catch (error) {
    console.error('Failed to get theme from localStorage:', error);
  }
  return { mode: 'automatic', name: 'default' };
}

/**
 * Set the theme in localStorage
 */
export async function setTheme(mode: ThemeMode, name: ThemeName): Promise<void> {
  try {
    const theme: StoredTheme = { mode, name };
    localStorage.setItem(THEME_STORAGE_KEY, JSON.stringify(theme));
  } catch (error) {
    console.error('Failed to set theme in localStorage:', error);
    throw error;
  }
}

/**
 * Get the system theme preference
 */
export async function getSystemTheme(): Promise<'day' | 'night'> {
  try {
    if (typeof window !== 'undefined' && window.matchMedia) {
      const isDark = window.matchMedia('(prefers-color-scheme: dark)').matches;
      return isDark ? 'night' : 'day';
    }
  } catch (error) {
    console.error('Failed to get system theme:', error);
  }
  return 'night';
}

/**
 * Reset theme to default
 */
export async function resetTheme(): Promise<void> {
  try {
    localStorage.removeItem(THEME_STORAGE_KEY);
  } catch (error) {
    console.error('Failed to reset theme:', error);
    throw error;
  }
}

File: src-tauri/src/theme.rs
//! Theme management module
//!
//! Handles theme persistence and system theme detection

use tauri::AppHandle;
use tauri_plugin_store::StoreExt;

const THEME_STORE_FILE: &str = "theme.json";
const THEME_MODE_KEY: &str = "mode";
const THEME_NAME_KEY: &str = "name";

#[derive(serde::Serialize, serde::Deserialize)]
pub struct ThemePreference {
    pub mode: String,
    pub name: String,
}

/// Get the saved theme preference
#[tauri::command]
pub async fn get_theme(app: AppHandle) -> Result<ThemePreference, String> {
    let store = app
        .store(THEME_STORE_FILE)
        .map_err(|e| format!("Failed to access theme store: {}", e))?;

    let mode = match store.get(THEME_MODE_KEY) {
        Some(value) => value.as_str().unwrap_or("automatic").to_string(),
        None => "automatic".to_string(),
    };

    let name = match store.get(THEME_NAME_KEY) {
        Some(value) => value.as_str().unwrap_or("default").to_string(),
        None => "default".to_string(),
    };

    Ok(ThemePreference { mode, name })
}

/// Save the theme preference
#[tauri::command]
pub async fn set_theme(app: AppHandle, mode: String, name: String) -> Result<(), String> {
    // Validate theme mode
    let valid_modes = ["day", "night", "automatic"];
    if !valid_modes.contains(&mode.as_str()) {
        return Err(format!(
            "Invalid theme mode: {}. Must be one of: day, night, automatic",
            mode
        ));
    }

    // Validate theme name
    let valid_names = ["default", "sky-blue", "cosmic-gold", "starry-christmas"];
    if !valid_names.contains(&name.as_str()) {
        return Err(format!(
            "Invalid theme name: {}. Must be one of: default, sky-blue, cosmic-gold, starry-christmas",
            name
        ));
    }

    let store = app
        .store(THEME_STORE_FILE)
        .map_err(|e| format!("Failed to access theme store: {}", e))?;

    store.set(THEME_MODE_KEY, serde_json::json!(mode));
    store.set(THEME_NAME_KEY, serde_json::json!(name));

    store
        .save()
        .map_err(|e| format!("Failed to save theme: {}", e))?;

    Ok(())
}

/// Get the current system theme (for automatic mode)
/// Returns "day" or "night" based on system appearance
#[tauri::command]
pub fn get_system_theme() -> String {
    // On macOS, we can use dark-light crate, but for simplicity
    // we'll return "night" as default and let the frontend handle it
    // via CSS media query prefers-color-scheme

    // For now, return a default that the frontend can override
    // The frontend will use window.matchMedia('(prefers-color-scheme: dark)')
    "night".to_string()
}

/// Reset theme to default (automatic)
#[tauri::command]
pub async fn reset_theme(app: AppHandle) -> Result<(), String> {
    set_theme(app, "automatic".to_string(), "default".to_string()).await
}
